src/main/scala/org/apache/spark/shuffle/ucx/UcxWorkerWrapper.scala

```diff
@@ -0,0 +1,134 @@
+
+import com.fasterxml.jackson.databind.util.ByteBufferBackedOutputStream
+import org.openucx.jucx.ucp.{UcpEndpoint, UcpEndpointParams, UcpRequest, UcpWorker}
+import org.openucx.jucx.{UcxCallback, UcxException, UcxUtils}
+import org.apache.spark.internal.Logging
+import org.apache.spark.shuffle.ucx.memory.MemoryPool
+
+class UcxSuccessOperationResult(stats: OperationStats) extends OperationResult {
+
+  override def getStatus: OperationStatus.Value = OperationStatus.SUCCESS
+
+  override def getError: TransportError = null
+
+  override def getStats: Option[OperationStats] = Some(stats)
+}
+
+class UcxFailureOperationResult(errorMsg: String) extends OperationResult {
+  override def getStatus: OperationStatus.Value = OperationStatus.FAILURE
+
+  override def getError: TransportError = new TransportError(errorMsg)
+
+  override def getStats: Option[OperationStats] = None
+}
+
+/**
+ * Worker per thread wrapper, that maintains connection and progress logic.
+ */
+class UcxWorkerWrapper(worker: UcpWorker, transport: UcxShuffleTransport, ucxConf: UcxShuffleConf,
+                       memoryPool: MemoryPool)
+  extends Closeable with Logging {
+
+  private final val connections = mutable.Map.empty[String, UcpEndpoint]
+  // Tag on which this worker would listen to
+  private val tag = Random.nextLong()
+  private val workerAddress = worker.getAddress
+
+  override def close(): Unit = {
+    connections.foreach{
+      case (_, endpoint) => endpoint.close()
+    }
+    connections.clear()
+    worker.close()
+  }
+
+  /**
+   * The only place for worker progress
+   */
+  private[ucx] def progress() {
+    if ((worker.progress() == 0) && ucxConf.useWakeup) {
+      worker.waitForEvents()
+      worker.progress()
+    }
+  }
+
+  private def getConnection(executorId: String): UcpEndpoint = {
+    val workerAdresses = transport.executorIdToAddress
+
+    if (!workerAdresses.contains(executorId)) {
+      // Block untill there's no worker address for this BlockManagerID
+      val startTime = System.currentTimeMillis()
+      val timeout = ucxConf.getTimeAsMs("spark.network.timeout", "100")
+      workerAdresses.synchronized {
+        while (workerAdresses.get(executorId) == null) {
+          workerAdresses.wait(timeout)
+          if (System.currentTimeMillis() - startTime > timeout) {
+            throw new UcxException(s"Didn't get worker address for $executorId during $timeout")
+          }
+        }
+      }
+    }
+
+    connections.getOrElseUpdate(executorId, {
+      logInfo(s"Worker from thread ${Thread.currentThread().getName} connecting to $executorId")
+      val endpointParams = new UcpEndpointParams()
+        .setUcpAddress(workerAdresses.get(executorId))
+     worker.newEndpoint(endpointParams)
+    })
+  }
+
+  def fetchBlockByBlockId(executorId: String, blockId: BlockId,
+                          resultBuffer: MemoryBlock, cb: OperationCallback): UcxRequest = {
+    val ep = getConnection(executorId)
+    workerAddress.rewind()
```

Comment:
`ByteBuffer` has a `put(byte[])` method also. Wonder if we can get the `byte[]` from `workerAddress` s.t. we don't need to rewind each time?

Reply:
Direct Byte buffer doesn't have `byte[]` method. We use direct byte buffer once to get worker address